### ISSUE TITLE
Service state Green and OK for return code of 255

### ISSUE THREAD

**Issue by originalrexconsulting** [NagiosEnterprises/nagioscore#144]
It seems that version 2.15 of "check_nrpe" returns error code 255 instead of error code 2 on a connection refused. Nagios server registers a return code of 255 as GREEN and OK and does not notify. 

This is really bad when you set up service dependencies so that all your NRPE-based service checks depend on NRPE. If the NRPE is down and "check_nrpe" is returning 255, Nagios is fine with all of that, and you end up with false negatives - and alerts that you may be missing.

**Comment by box293** [comment]
This needs to be reported in the NRPE GitHub project please.

https://github.com/NagiosEnterprises/nrpe

FYI NRPE v3 is now released. Can you test with NRPE v3 and see if your issue is resolved. There will no longer be any released for version 2.x of NRPE.

Here are some guides:

Installation:
https://support.nagios.com/kb/article.php?id=515

An upgrade guide:
https://support.nagios.com/kb/article.php?id=520

**Comment by originalrexconsulting** [comment]
Yeah I thought about reporting it as an NRPE problem but I also see a problem with Nagios server and how it handles plugin return codes. Any plugin returning other than a 1, 2, or 3 is apparently OK, which does not seem right to me. It should be that if and only if a plugin returns zero to nagios will the check return OK. And note it is "check_nrpe" not "nrpe" that is returning 255, which I believe is another separate issue than the one where Nagios is OK with a non-zero return code.

What adds to this problem is that while in this mode where the plugin is returning 255, the service check will light up Green on the Nagios core status.cgi screens. And in the case that it's NRPE that we're checking and check_nrpe is returning 255, it's causing dependent services to light up Green too. So we have a server that does not have NRPE running but Nagios shows all its services as Green.

**Comment by ericloyd** [comment]
Are you SURE that the plugin is not returning a zero result, but passing along output that says "returned code 255"?

**Comment by box293** [comment]
Good question Eric.
Can we get a command line example of the plugin being executed along with the output and exit code:
```
./check_xxx --options blah
This is the output

echo $?
A number should be returned by that command
```

**Comment by originalrexconsulting** [comment]
$ /usr/local/nagios/bin/check_nrpe -p 5666 -n  -H obf1
connect to address 123.123.123.123 port 5666: Connection refused
connect to host obf1 port 5666: Connection refused[m70209@wlpd069 ~]$ echo $?
255

Weird how it does not carriage return/line feed after the "Connection refuse".

Note to run this test, I just turned off the nrpe daemon on the host "obf1" and then ran "check_nrpe" against it.

Come to think of it, I am not exactly sure where this check_nrpe came from. I believe the standard 2.15 source code, compiled on Red Hat Enterprise Linux Server release 6.7. I can test a fresh compile when I can get a few spare minutes.

**Comment by box293** [comment]
OK great that helped.

Good news is that you've pinpointed the error is in NRPE.

The other great bit of news is that this has been fixed in NRPE v3.0

Furthmore, your comment about "Weird how it does not carriage return/line feed after the "Connection refused" has also been resolved in NRPE v3.0.

Here is NRPE v2.15 replicating your issue:
```
./check_nrpe -p 5666 -n -H rhel02
connect to address 10.25.13.53 port 5666: Connection refused
connect to host rhel02 port 5666: Connection refused[root@xitest libexec]# 

echo $?
255
```
Here is NRPE v3.0 showing the correct behaviour:
```
./check_nrpe -H rhel02
connect to address 10.25.13.53 port 5666: Connection refused
connect to host rhel02 port 5666: Connection refused

echo $?
2
```
When I say it's fixed in NRPE v3, specifically it's resolved in the check_nrpe binary. To fix the issue on your Nagios server you only need to install the NPRE v3 check_nrpe binary, which is clearly explained at the bottom of this guide (install guide):
https://support.nagios.com/kb/article.php?id=515

Here is the documentation for upgrading NRPE v3:
https://support.nagios.com/kb/article.php?id=520

**Comment by jfrickson** [comment]
Sounds like 3.0 resolved the issues. Closing.
